### ISSUE TITLE
Capture with `k` workers, replay with `j`, `k ≠ j`

### ISSUE THREAD

**Issue by utaal** [frankmcsherry/timely-dataflow#81]
`.capture()`-d streams may be sent over sockets to a separate computation.

It's currently unclear how to replay those streams with a different number of workers compared to the source computation.

For example, replaying a 1-worker stream in a j-worker stream, might look like this - but will not work.
```
                let one_stream: Stream<_, Thing> = if index == 0 {
                    one_stream.lock().unwrap().take().unwrap().replay_into(scope)
                } else {
                    vec![].to_stream(scope)
                };
```

The broken progress behaviour seems due to `.to_stream()` expecting to exist on every worker, https://github.com/frankmcsherry/timely-dataflow/blob/eae63eb2063a6e1d3da70c781c3e879a5ccd7d93/src/dataflow/operators/generic/operator.rs#L554

We may just need a source that does not advertise initial capabilities? I'm not sure this checks out with the progress tracking math.

**Comment by ms705** [comment]
Ah, this is good too know about -- I was about to implement a similar paradigm in a situation where _n_ clients talk to _m_ workers (for _n_ != _m_).

I guess it would be useful to have a "multiplex" paradigm, where one dataflow is split into many (for "many" = _m_ in this case). I guess maybe `partition` can be massaged into serving this purpose?

**Comment by frankmcsherry** [comment]
My plan (maybe a bad plan) was to put together a source that could play back arbitrary numbers of sources, from zero to lots, at each worker. The logic isn't too complicated, I think; you still initialize with as many capabilities as there are workers *in the current computation*, and then have each worker maintain the capabilities needed for its collective streams. This might mean immediately dropping the capability if it has no streams, and it might mean a `MutableAntichain` to track the lower bounds if it has multiple streams.

I'm not sure the best way to do this generally, for `replay` as well as other sources of inputs (I'm looking at reading Kafka queues right now, for example), but maybe doing it once we can look at the result and see if we like it?

**Comment by frankmcsherry** [comment]
More generally, if you want to have multiple "proxy workers" managed by one worker, it should be pretty easy: each worker in most *source* operators apply a constructor where they get a capability, and produce a closure that acts on the output handle.

It seems pretty easy to rig a single worker to manage multiple workers *for sources*: in initialization you clone your capability as many times as you have workers to manage, invoke each of their constructors and assemble their closures in a `Vec`. When you are invoked with each handle, you call each of your stashed closures on the output handle.

This almost works for general operators, except that we would need to demux their inputs. It is totally worth thinking that out though, as proxying operators is a fair way to think about doing elasticity.

**Comment by utaal** [comment]
I'm taking a look at the current .replay_into() implementation and that seems to only publish to `get_internal_summary` the capabilities from its own captured stream (without "multiplying" by the number of workers, similarly to what `source` does): is it because the captured progress tracking information already takes into account the number of works?

**Comment by frankmcsherry** [comment]
It doesn't look like it, though paging all this code back in is a bit .. right. 

It looks like the first entry of the list gets set in `CaptureOperator`'s `set_external_summary` method. This is what the replay is going to pop open and announce when it starts up. 

```rust
    // we need to set the initial value of the frontier
    fn set_external_summary(&mut self, _: Vec<Vec<Antichain<T::Summary>>>, counts: &mut [ChangeBatch<T>]) {
        let mut map = counts[0].clone();
        map.update(Default::default(), -1);
        if !map.is_empty() {
            self.events.push(Event::Progress(map.into_inner()));
        }
        counts[0].clear();
    }
```

So it seems fairly likely that when starting up, there should be a bit of a mess with multiple workers. I feel like we have tested this in the past, so either (i) it works for mysterious reasons, (ii) it totally doesn't work but we haven't tested it well enough, or (iii) there is a race ("doesn't work") that we haven't seen yet.

I think, if our understanding is right, that starting this with multiple workers should end up in a state where there are negative counts for capabilities for each worker (each initially says "only one capability" to its hosting subgraph, who then hears about multiple such capabilities dropped).

**Comment by frankmcsherry** [comment]
My mistake; where the loading actually happens is 

```rust
impl<T:Timestamp, D: Data, P: EventPusher<T, D>> CaptureOperator<T, D, P> {
    fn new(input: PullCounter<T, D>, mut events: P) -> CaptureOperator<T, D, P> {
        events.push(Event::Progress(vec![(Default::default(), 1)]));
        CaptureOperator {
            input: input,
            events: events,
        }
    }
}
```

I *think* this `1` should be a `peers`, which we don't have here but could easily get. The statement that the `Capture` operator's `set_external_summary` makes is about the first transition from the default capability, as supplied by the world immediately outside `capture`.

**Comment by frankmcsherry** [comment]
Here is a proposal for the capture/replay protocol that I think simplifies the `m != n` case greatly:

1. The first thing that goes into the stream is an `Event::Start`, which currently has no data but should be augmented with "number of peers in source stream". This should always be the first token in the stream, and it tells any replayer what multiplificity to use for the capability when initializing (this number is important because it is "how many capability *drops* should you expect to see).

2. Each stream then, in `set_external_summary` indicates a transition from `Default::default()` to whatever external capability they have presented at them. It may be a no-op, which is fine.

3. On replay, the replayer expects a `Event::Start` as the first event, uses its multiplicity, and then ignores all other `Event::Start` events.

I *believe* that this would allow us to arbitrarily interleave any captured streams into replayed streams. That is, if each of the captured streams were fed into some smaller number of `mpsc` channels, the receiver would (i) correctly identify the number of peers initially, and (ii) play out a stream that always retained capabilities needed for any message send.

If the replayer has *no* input streams, in this protocol they would have to be quiet and neither express initial capabilities nor ever send any data, which might trigger timely dataflow's "you probably have a bug" logic, but seems like it would conform to the progress protocol.

**Comment by utaal** [comment]
So, I'm printing out the initialisation (`get_internal_summary`) for EventReader, and it looks like it's only advertising `(Default::default(), 1)` which should probably be `(Default::default(), peers)` and seems to confirm `CaptureOperator` should advertise `(Default::default(), peers)` in `fn new`.

**Comment by frankmcsherry** [comment]
I just pushed a branch that *should* fix this.  https://github.com/frankmcsherry/timely-dataflow/tree/capture_replay

The idea here is that you now do `replay_into` from iterators of sources. You could have this be `Some(source)` for your old source, but you would no longer be able to do `source.replay_into(...)`. There is a bit of a pain here in that both of these implementations are for `I: Trait` for two different traits, and there is no way to know that they are not conflicting implementations. It seems hard to do both at the same time. Let's discuss whether it is better to break things in the name of generality.

So, what you can do now is, with `examples/capture_send.rs` more or less unchanged (changed to look at args and maybe use multiple workers) is to write `examples/capture_recv.rs` as

```rust
extern crate timely;

use std::net::TcpListener;
use timely::dataflow::operators::Inspect;
use timely::dataflow::operators::capture::{EventReader, Replay};

fn main() {
    timely::execute_from_args(std::env::args(), |worker| {

        let source_peers = std::env::args().nth(1).unwrap().parse::<usize>().unwrap();

        let mut listeners = Vec::new();

        for source_peer in 0 .. source_peers {
            if source_peer % worker.peers() == worker.index() {
                let addr = format!("127.0.0.1:{}", 8000 + source_peer);
                listeners.push(TcpListener::bind(addr).unwrap());
            }
        }

        let mut replayers = Vec::new();
        for listener in listeners {
            let recv = listener.incoming().next().unwrap().unwrap();
            replayers.push(EventReader::<_,u64,_>::new(recv));
        }

        worker.dataflow::<u64,_,_>(|scope| {
            replayers
                .replay_into(scope)
                .inspect(|x| println!("replayed: {:?}", x));
        })
    }).unwrap(); // asserts error-free execution
}
```

It takes an argument indicating the number of source peers, and then has the workers divide out responsibility among them. Once that it done, they each assemble a vector of `EventReader` types, from which they can `replay_into`.

I tried it with five source workers and three receive workers, and it "worked fine" meaning no errors and proper termination. There is no test that ensures that the progress protocol didn't violate any invariants, though.

**Comment by utaal** [comment]
Thanks! Just to make sure: the `replayers` Vec is for the replayers to be replayed in the _current_ worker, right?

**Comment by ms705** [comment]
@frankmcsherry This looks good!

One observation relevant for multi-client-multi-server use cases: this does statically bind source peers to specific workers, rather than multiplexing the stream (e.g., doing a round-robin over the workers at each source peer). In the limit, this doesn't seem concerned: if there are many more clients than servers, the work should balance. With small numbers (e.g., 5 source peers, 3 workers), however, you would end up with some imbalance in the work done by each worker, correct?

**Comment by frankmcsherry** [comment]
@utaal : yes, that should work, I think. It is a bit like `source_peers = 0` in the example above. You could also have worker zero play out all streams (not just one); it is *supposed* to work (whether the bugs are shaken out of new code is a diff questions).

@ms705 : This does have a static binding, and we could try to think of a more balanced approach. The issue with that here is that each worker needs to track and report progress, so if each is working through the same set of streams, each progress event in each stream will need to be replicated by workers but each data event will need to be partitioned between. This has the aesthetically unappealing behavior that if you put a sequence of these in place, each step in the sequence multiplies the progress traffic by `#workers`.

It seems like a reasonable think to ponder. Maybe another way to do it would be to have `capture_into` be able to capture into finer granularity (e.g. 256 streams, even if greater than workers) so that round-robin replay is more balanced?

**Comment by frankmcsherry** [comment]
There is a PR over at #84 that should fix this issue, the bug detected along the way, and introduce a bit more documentation on what needs to be true about the communication between `capture` and `replay`.

**Comment by frankmcsherry** [comment]
This just got fixed by #84, but I don't type guds so close by hands.
